src/main/java/com/rabbitmq/jms/client/RMQSession.java

```diff
@@ -446,6 +457,13 @@ public void rollback() throws JMSException {
             try {
                 // rollback the RabbitMQ transaction which may cause some messages to become unacknowledged
                 this.channel.txRollback();
+                if (this.commitNackOnRollback && this.uncommittedMessageTags.size() > 0) {
+                    for (Long dtag : this.uncommittedMessageTags) {
+                        this.channel.basicNack(dtag, false, false);
+                    }
+                    this.channel.txCommit();
+	        }
+                this.uncommittedMessageTags.clear();
                 // requeue all unacknowledged messages (not automatically done by RabbitMQ)
                 this.channel.basicRecover(true); // requeue
```

Comment:
In theory this call should not be necessary when the feature is enabled because all the unack-ed messages are supposed to have been ack-ed. Nevertheless I think it can stay there: this makes sure that no unack-ed message never goes back to the server.